---
title: "Fst and Bartlett Test"
author: "AJM & CLM"
date: "5/8/2019"
output: github_document
editor_options: 
  chunk_output_type: console
---
Rmd #2

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library("adegenet") #For storing genotype data
library(hierfstat) #For calculating pairwise Fst
library(ggplot2) #For plotting
library(reshape2) #For plotting
library(plyr)
library("cowplot") #For plotting manuscript figs
library(readr)
```
```{r set directory to where radiator output is}
setwd("~/github/Erigenia-Analyses/a118_radiator_genomic_converter_20190508a1444")
```

begin at section named Combined in Silliman's Ostrea tutorial, load previously created adegenet genind object and hierfstat object, made from radiator from vcf

```{r read in genind}
radiator_genind_a <- read_rds("/Volumes/Thunderbolt/Users/PostDoc/Github/Erigenia-Analyses/a118_radiator_genomic_converter_20190508a1444/radiator_genind_20190508@1445.RData")
```
```{r read in heirfstat}
radiator_hf_a <- read.fstat("/Volumes/Thunderbolt/Users/PostDoc/Github/Erigenia-Analyses/a118_radiator_genomic_converter_20190508a1444/radiator_data_20190508a1444.dat", na.s = c("0","00","000","0000","00000","000000","NA"))
```
```{r population order for later}
pop_order <- c("BTP","CCP","MD","PB","RCSPWR","RSSP","SH", "SRCNA", "YF")
```


This takes a long time (~30 minutes)
```{r calculate fst}
fst.mat<-genet.dist(radiator_hf_a, method = "WC84")
```
```{r}
gindF.fst.mat.tri <- as.matrix(fst.mat)
colnames(gindF.fst.mat.tri)<- pop_order
rownames(gindF.fst.mat.tri)<- pop_order
write.table(gindF.fst.mat.tri, file="ErigeniaFstWC84.pwfst", row.names=T,col.names=T)
```
This is so we can arrange these east to west
```{r}
newmelted<-read.csv("/Volumes/Thunderbolt/Users/PostDoc/Desktop/newmelted.csv", sep=",", header=TRUE)
```
```{r}
melted <- melt(newmelted, na.rm = TRUE)
```
```{r}
summary(melted$value)
```

FIGURE FOR PUB
plotting using ggplot2
```{r}
library(cowplot)
```
```{r figure for fst}
combined <- ggplot(data = melted, aes(Var2, Var1, fill = value))+ geom_tile(color = "white")+ 
  scale_fill_gradient(low = "white", high = "red", name="FST")  +
  ggtitle(expression(atop("Pairwise FST, WC (1984), Combined", atop(italic("N = 124, L = 9,243"), ""))))+
  labs( x = "Sampling Site", y = "Sampling Site") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, size = 12, hjust = 1),axis.text.y = element_text(size = 13)) + 
  coord_fixed()
combined
```

```{r saving tiff}
save_plot("Erigenia_Fst_figure.pdf", plot = combined, ncol = 1, nrow = 1, base_height = 4, base_aspect_ratio = 1.1, base_width = NULL, cols = NULL,rows = NULL, dpi = 600)
```

#not super informative figures, but a good check of data
relabel individuals by their geographic region
```{r}
hf_a_reg<- radiator_hf_a
regions<- c("West", "West", "MD", "East", "West", "West", "East", "West", "East")
hf_a_reg$Pop <- mapvalues(hf_a_reg$Pop, from =c(1,2,3,4,5,6,7,8,9), to = regions)
```
calculate fst matrix for regions
```{r}
fst.mat.reg <- genet.dist(hf_a_reg, method="WC84")
```
```{r}
EW <- c("West", "MD", "East")
EW.fst.mat.tri <- as.matrix(fst.mat.reg)
EW.fst.mat.tri <- EW.fst.mat.tri[EW,EW]
```
```{r}
write.table(EW.fst.mat.tri, file="ErigeniaFstWC84Reg.pwfst", row.names=T, col.names=T)
EW.fst.mat.tri[lower.tri(fst.mat.reg, diag=TRUE)] <-NA
melted.reg <- melt(EW.fst.mat.tri, na.rm=TRUE)
```

```{r}
par(mfrow=c(2,1))
ggplot(data = melted.reg, aes(Var2, Var1, fill = value))+ geom_tile(color = "white")+ 
  scale_fill_gradient(low = "white", high = "red", name="FST")  +
  ggtitle(expression(atop("Pairwise FST, WC (1984), Regions", atop(italic("N = 124, L = 9,243"), ""))))+
  labs( x = "Region", y = "Region") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, size = 12, hjust = 1),axis.text.y = element_text(size = 13)) + 
  coord_fixed()
```
#need to remove MD. prob can't repeat above with 2 regions. 

#MD removed! see end of Erigenia_Python_perSilliman.Rmd, made hierfstat & genlight (genind impossible.)
Below is a rerun of much of above code with MD individuals
```{r}
setwd("~/github/Erigenia-Analyses/a166_radiator_genomic_converter_20190513a1509")
```
```{r}
radiator_genlight_a_noMD <- read_rds("/Volumes/Thunderbolt/Users/PostDoc/Github/Erigenia-Analyses/a166_radiator_genomic_converter_20190513a1509/radiator_genlight_20190513a1614.RData")
```
```{r}
radiator_hf_a_noMD <- read.fstat("/Volumes/Thunderbolt/Users/PostDoc/Github/Erigenia-Analyses/a166_radiator_genomic_converter_20190513a1509/radiator_data_20190513a1509.dat", na.s = c("0","00","000","0000","00000","000000","NA"))
```
```{r}
pop_order_noMD <- c("BTP","CCP","PB","RCSPWR","RSSP","SH", "SRCNA", "YF")
```
This takes a long time (~24 hrs for all snp's in Erigenia)
```{r}
noMD.fst.mat<-genet.dist(radiator_hf_a_noMD, method = "WC84")
```
```{r save the thing cause it takes forever to make}
save(noMD.fst.mat, file="noMD.fst.mat.RData")
```

```{r}
noMD.gindF.fst.mat.tri <- as.matrix(noMD.fst.mat)
colnames(noMD.gindF.fst.mat.tri)<- pop_order_noMD
rownames(noMD.gindF.fst.mat.tri)<- pop_order_noMD
write.table(noMD.gindF.fst.mat.tri, file="ErigeniaFstWC84.NoMD.pwfst", row.names=T,col.names=T)
```
             BTP       CCP        PB    RCSPWR      RSSP        SH     SRCNA        YF
BTP    0.0000000 0.3072230 0.5573158 0.2177214 0.2288916 0.6100386 0.2217568 0.5040910
CCP    0.3072230 0.0000000 0.6111765 0.2921333 0.2962880 0.6748338 0.2919878 0.5489660
PB     0.5573158 0.6111765 0.0000000 0.4682329 0.4692480 0.3247984 0.4538422 0.1967257
RCSPWR 0.2177214 0.2921333 0.4682329 0.0000000 0.1639937 0.4988519 0.1670948 0.4118465
RSSP   0.2288916 0.2962880 0.4692480 0.1639937 0.0000000 0.4947547 0.1871296 0.4331413
SH     0.6100386 0.6748338 0.3247984 0.4988519 0.4947547 0.0000000 0.4816955 0.2401579
SRCNA  0.2217568 0.2919878 0.4538422 0.1670948 0.1871296 0.4816955 0.0000000 0.4069650
YF     0.5040910 0.5489660 0.1967257 0.4118465 0.4331413 0.2401579 0.4069650 0.0000000



use below read in edited melted csv, if needed. (add a,b,c, etc. as needed to orient pops W to E)
```{r}
newmelted_noMD<-read.csv("/Volumes/Thunderbolt/Users/PostDoc/Desktop/newmelted_noMD.csv", sep=",", header=TRUE)
```
melt the matrix made above
```{r}
melted_noMD <- melt(noMD.gindF.fst.mat.tri, na.rm = TRUE)
```
```{r}
summary(newmelted_noMD$value)
```
plotting using ggplot2
```{r}
combined_noMD <- ggplot(data = newmelted_noMD, aes(Var2, Var1, fill = value))+ geom_tile(color = "white")+ 
  scale_fill_gradient(low = "white", high = "red", name="FST")  +
  ggtitle(expression(atop("Pairwise FST, WC (1984), Combined", atop(italic("N = 120, L = 220,427"), ""))))+
  labs( x = "Sampling Site", y = "Sampling Site") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, size = 12, hjust = 1),axis.text.y = element_text(size = 13)) + 
  coord_fixed()
combined_noMD
```
save fig
```{r saving no MD tiff}
save_plot("Erigenia_Fst_figure_noMD.pdf", plot = combined_noMD, ncol = 1, nrow = 1, base_height = 4, base_aspect_ratio = 1.1, base_width = NULL, cols = NULL,rows = NULL, dpi = 600)
```


#calculating regional fst's for reporting only (no graphs)
relabel individuals by their geographic region
#No figures for this
```{r}
hf_noMD_reg <- radiator_hf_a_noMD
regions<- c("West", "West", "East", "West", "West", "East", "West", "East")
hf_noMD_reg$Pop <- mapvalues(hf_noMD_reg$Pop, from =c(1,2,3,4,5,6,7,8), to = regions)
```
```{r}
noMD.fst.mat.reg <- genet.dist(hf_noMD_reg, method="WC84")
```

```{r}
EW <- c("West", "East")
EW.fst.mat.tri <- as.matrix(noMD.fst.mat.reg)
EW.fst.mat.tri <- EW.fst.mat.tri[EW,EW]
```
```{r}
write.table(EW.fst.mat.tri, file="ErigeniaFstWC84EW.pwfst", row.names=T, col.names=T)
EW.fst.mat.tri[lower.tri(noMD.fst.mat.reg, diag=TRUE)] <-NA
EW.fst.mat.tri
```
output
      West      East
West   NA 0.3768538
East   NA        NA

Not super informative and no check on if these were grouped correctly

#Bartlett Test

need a genind to do bartlett test of observed vs expected heterozygosity, convert no md vcf to genind using VCFtools (manual/site: https://vcftools.github.io/man_latest.html)

```{bash}
vcftools --vcf erigeniaC.vcf --recode --recode-INFO-all --min-alleles 2 --max-alleles 2 --max-missing 0.75 --keep indiv.txt --out filtered_noMD
```
VCFtools - 0.1.16
(C) Adam Auton and Anthony Marcketta 2009

Parameters as interpreted:
	--vcf erigeniaC.vcf
	--keep indiv.txt
	--recode-INFO-all
	--max-alleles 2
	--min-alleles 2
	--max-missing 0.75
	--out filtered_noMD
	--recode

Keeping individuals in 'keep' list
After filtering, kept 120 out of 124 Individuals
Outputting VCF file...
After filtering, kept 22624 out of a possible 439647 Sites
Run Time = 22.00 seconds

```{r}
Erigenia_filtered_noMD.recode.vcf <- read.vcfR("/Volumes/G-DRIVEUSB/Erigenia/ErigeniaPEdataandruns/ErigeniaC/erigeniaC_outfiles/filtered_noMD.recode.vcf")
                                               
```
```{r}
library(vcfR)
```
```{r}
Erigenia_filtered_noMD.recode.genind <- vcfR2genind(Erigenia_filtered_noMD.recode.vcf)
```
```{r}
save(Erigenia_filtered_noMD.recode.genind,file="Erigenia_filtered_noMD.recode.genind.Rdata")
```
assigning populations to samples, assuming samples are in same order as in vcf.
```{r}
pop(Erigenia_filtered_noMD.recode.genind)<-as.factor(c("PB","PB","PB","PB","PB","PB","PB","PB","PB","PB","PB","PB","PB","PB","PB","PB","PB","PB","PB","PB","YF","YF","YF","YF","YF","YF","YF","YF","YF","YF","YF","YF","YF","YF","YF","SH","SH","SH","SH","SH","SH","SH","SH","SH","SH","CCP","CCP","CCP","CCP","CCP","CCP","CCP","CCP","CCP","CCP","CCP","CCP","CCP","CCP","CCP","BTP","BTP","BTP","BTP","BTP","BTP","BTP","BTP","BTP","BTP","BTP","BTP","BTP","BTP","BTP","RSSP","RSSP","RSSP","RSSP","RSSP","RSSP","RSSP","RSSP","RSSP","RSSP","RSSP","RSSP","RSSP","RSSP","RSSP","RCSPWR","RCSPWR","RCSPWR","RCSPWR","RCSPWR","RCSPWR","RCSPWR","RCSPWR","RCSPWR","RCSPWR","RCSPWR","RCSPWR","RCSPWR","RCSPWR","RCSPWR","SRCNA","SRCNA","SRCNA","SRCNA","SRCNA","SRCNA","SRCNA","SRCNA","SRCNA","SRCNA","SRCNA","SRCNA","SRCNA","SRCNA","SRCNA"))
popNames(Erigenia_filtered_noMD.recode.genind)
```

use tut here https://popgen.nescent.org/StartSNP.html to get Ho/Hexp, Bartlett test, etc. for all of Erigenia in PA.
```{r}
library("pegas")
```
```{r}
ErigeniaHet <- summary(Erigenia_filtered_noMD.recode.genind)
ErigeniaHet
```

```{r saving het}
save(ErigeniaHet,file="ErigeniaHet.Rdata")
```

```{r}
names(ErigeniaHet)
```

SAVING
save fig

```{r}
tiff(filename="Erigenia_Hob_plot.tiff", width = 4, height = 4, units = 'in', res = 300)
plot(ErigeniaHet$Hobs, xlab="Locus number", ylab="Observed heterozygosity", main="Observed heterozygosity per locus")
dev.off()
```


```{r}
tiff(filename="Erigenia_Ho_He_plot.tiff", width = 4, height = 4, units = 'in', res = 300)
plot(ErigeniaHet$Hobs, ErigeniaHet$Hexp, xlab="Observed heterozygosity", ylab="Expected heterozygosity", main="He as a function of Ho, per locus")
dev.off()
```








```{r}
bartlett.test(list(ErigeniaHet$Hexp, ErigeniaHet$Hobs))
```
	Bartlett test of homogeneity of variances
Bartlett's K-squared = 24950, df = 1, p-value < 2.2e-16
wow: across PA, Ho is very low, statistically significantly so.

Now, separate pops and repeat Bartlett for each.
Use seppop in adegenet to split the genind by population. use pop=NULL to let it use the pop slot in the big genind, makes a list.
```{r}
temp <- seppop(Erigenia_filtered_noMD.recode.genind)
```
make individual geninds from the list
```{r}
PB <- temp$PB
BTP <- temp$BTP
CCP <- temp$CCP
RCSPWR <- temp$RCSPWR
RSSP <- temp$RSSP
SH <- temp$SH
SRCNA <- temp$SRCNA
YF <- temp$YF
```

each pop can now be called! do bartletts for each.
```{r}
PBHet <- summary(PB)
bartlett.test(list(PBHet$Hexp, PBHet$Hobs))
```
Bartlett's K-squared = 7552.4, df = 1, p-value < 2.2e-16
woo!

```{r}
BTPHet <- summary(BTP)
bartlett.test(list(BTPHet$Hexp, BTPHet$Hobs))
```
Bartlett's K-squared = 5504, df = 1, p-value < 2.2e-16

```{r}
CCPHet <- summary(CCP)
bartlett.test(list(CCPHet$Hexp, CCPHet$Hobs))
```
Bartlett's K-squared = 7075.6, df = 1, p-value < 2.2e-16

```{r}
RCSPWRHet <- summary(RCSPWR)
bartlett.test(list(RCSPWRHet$Hexp, RCSPWRHet$Hobs))
```
Bartlett's K-squared = 6327.5, df = 1, p-value < 2.2e-16

```{r}
RSSPHet <- summary(RSSP)
bartlett.test(list(RSSPHet$Hexp, RSSPHet$Hobs))
```
Bartlett's K-squared = 4835.7, df = 1, p-value < 2.2e-16

```{r}
SHHet <- summary(SH)
bartlett.test(list(SHHet$Hexp, SHHet$Hobs))
```
Bartlett's K-squared = 1318.8, df = 1, p-value < 2.2e-16

```{r}
SRCNAHet<- summary(SRCNA)
bartlett.test(list(SRCNAHet$Hexp, SHHet$Hobs))
```
Bartlett's K-squared = 11485, df = 1, p-value < 2.2e-16

```{r}
YFHet <- summary(YF)
bartlett.test(list(YFHet$Hexp, SHHet$Hobs))
```
Bartlett's K-squared = 6642.3, df = 1, p-value < 2.2e-16

