---
title: "Erigenia_Fst_perSilliman"
author: "AJM & CLM"
date: "5/8/2019"
output: github_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library("adegenet") #For storing genotype data
library(hierfstat) #For calculating pairwise Fst
library(ggplot2) #For plotting
library(reshape2) #For plotting
library(plyr)
library("cowplot") #For plotting manuscript figs
library(readr)
```
```{r}
setwd("~/github/Erigenia-Analyses/118_radiator_genomic_converter_20190508@1444")
```

begin at section named Combined, load previously created Adegenet genind object and hierfstat object, made from radiator from vcf

```{r}
radiator_genind_a <- read_rds("/Volumes/Thunderbolt/Users/PostDoc/Github/Erigenia-Analyses/118_radiator_genomic_converter_20190508@1444/radiator_genind_20190508@1445.RData")
```
```{r}
radiator_hf_a <- read.fstat("/Volumes/Thunderbolt/Users/PostDoc/Github/Erigenia-Analyses/a118_radiator_genomic_converter_20190508a1444/radiator_data_20190508a1444.dat", na.s = c("0","00","000","0000","00000","000000","NA"))
```
