---
title: "Erigenia_vcf_to_genlight"
author: "CLM"
date: "8/3/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This file converts our vcf file to a genlight file for use in downstream analyses

```{r library}
library(adegenet)
library(pegas)
library(vcfR)
library(dartR)
```

```{r}
setwd("~/Documents/RStudio/Erigenia/Erigenia_Analyses/vcf_tools")
```

```{r read in vcf with vcfR}
vcf<-read.vcfR("erigeniaC_noMD_filtered.recode_14350_118.vcf")
```


```{r convert}
new_Erigenia_gl <- vcfR2genlight(vcf)
```
  
#checking genlight
```{r}
indNames(new_Erigenia_gl)
```
```{r}
levels(pop(new_Erigenia_gl))
```

```{r}
glSum(new_Erigenia_gl, alleleAsUnit = TRUE, useC = FALSE)
glNA(new_Erigenia_gl, alleleAsUnit = TRUE)
glMean(new_Erigenia_gl, alleleAsUnit = TRUE)
glVar(new_Erigenia_gl, alleleAsUnit = TRUE)
```
need to assign pops 


```{r assigning pops}
pop(new_Erigenia_gl) <- as.factor(c("PeachBottom","PeachBottom","PeachBottom","PeachBottom","PeachBottom","PeachBottom","PeachBottom","PeachBottom","PeachBottom","PeachBottom","PeachBottom","PeachBottom","PeachBottom","PeachBottom","PeachBottom","PeachBottom","PeachBottom","PeachBottom","PeachBottom","PeachBottom","YorkFurnace","YorkFurnace","YorkFurnace","YorkFurnace","YorkFurnace","YorkFurnace","YorkFurnace","YorkFurnace","YorkFurnace","YorkFurnace","YorkFurnace","YorkFurnace","YorkFurnace","YorkFurnace","SafeHarbor","SafeHarbor","SafeHarbor","SafeHarbor","SafeHarbor","SafeHarbor","SafeHarbor","SafeHarbor","SafeHarbor","RaccoonCreek","RaccoonCreek","RaccoonCreek","RaccoonCreek","RaccoonCreek","RaccoonCreek","RaccoonCreek","RaccoonCreek","RaccoonCreek","RaccoonCreek","RaccoonCreek","RaccoonCreek","RaccoonCreek","RaccoonCreek","RaccoonCreek","CedarCreek","CedarCreek","CedarCreek","CedarCreek","CedarCreek","CedarCreek","CedarCreek","CedarCreek","CedarCreek","CedarCreek","CedarCreek","CedarCreek","CedarCreek","CedarCreek","CedarCreek","BraddockTrail","BraddockTrail","BraddockTrail","BraddockTrail","BraddockTrail","BraddockTrail","BraddockTrail","BraddockTrail","BraddockTrail","BraddockTrail","BraddockTrail","BraddockTrail","BraddockTrail","BraddockTrail","BraddockTrail","RyersonStation","RyersonStation","RyersonStation","RyersonStation","RyersonStation","RyersonStation","RyersonStation","RyersonStation","RyersonStation","RyersonStation","RyersonStation","RyersonStation","RyersonStation","RyersonStation","RyersonStation","SlipperyRockCreek","SlipperyRockCreek","SlipperyRockCreek","SlipperyRockCreek","SlipperyRockCreek","SlipperyRockCreek","SlipperyRockCreek","SlipperyRockCreek","SlipperyRockCreek","SlipperyRockCreek","SlipperyRockCreek","SlipperyRockCreek","SlipperyRockCreek","SlipperyRockCreek","SlipperyRockCreek"))
popNames(new_Erigenia_gl)
```
https://grunwaldlab.github.io/Population_Genetics_in_R/analysis_of_genome.html

alternatively, we could have used code like this 
assign pops to pop slot in genlight 
```{r}
#pop<-Erigenia.strata_noMD$LOCATION
```

#Doing Analyses

##DAPC
Helpful links: 
https://grunwaldlab.github.io/Population_Genetics_in_R/DAPC.html
http://adegenet.r-forge.r-project.org/files/tutorial-dapc.pdf
 -Code to explore DAPC objects and assignments
```{r}
grp<-find.clusters(new_Erigenia_gl, max.n.clust=8, n.iter= 100)
```
retained 120 PCs
chose K=8 first time, second run 6 retained based on BIC (lumps RCSPWR and BTP/ YF and SH)
Decided to do 8 because of labeling and downstream clarity

```{r}
grp
```
looks good, this is what we assigned

```{r}
dapc1<- dapc(new_Erigenia_gl, grp$grp)
```
120 pcs retained
retained all discriminate fucntions- 8 

```{r}
save(dapc1, file="dapc1.RData")
```

```{r}
dapc1$assign
```
perfect assignemnt 
```{r}
dapc1$posterior
```

```{r}
loadingplot(dapc1$var.contr)
```
```{r}
summary(dapc1)
```
prior and post same grouping

```{r}
assignplot(dapc1)
```
dang, perfect high prob assignment to origional pop



```{r save fig}
myCol <- c("burlywood4","darkviolet","green","orange","red","dodgerblue3","black","gold2")
#pdf("DAPC.pdf", width = 4, height = 4)
scatter(dapc1, posi.da="bottomright", bg="white",
pch=17:22, cstar=0, col=myCol, scree.pca=TRUE,
posi.pca="bottomleft")
#dev.off()
```
Where are the individual points?


Basic=scatter.dapc(Erigenia.dapc); Can add more arguments  to above code and change other features

```{r}
scatter(dapc1, scree.da=FALSE, bg="white", pch=20, cell=0, cstar=0, col=myCol, solid=.4,
cex=3,clab=0, leg=TRUE, txt.leg=paste("Cluster",1:8))
```


```{r}
scatter(dapc1, col=myCol, scree.da=FALSE,
cell=1.5, cex=2, bg="white",cstar=0)
```



Explore other PCs and optimize
------------- Investigating number of PCs with alpha values etc
http://adegenet.r-forge.r-project.org/files/tutorial-dapc.pdf pg 37

```{r}
temp <- a.score(dapc1)
names(temp)
temp<- optim.a.score(dapc1)
```
number retained pcs ~12
```{r}
temp$pop.score
temp$mean
```

```{r}
dapc2 <- dapc(new_Erigenia_gl)
```
n.pca=12, n.da=3 

```{r}
temp2 <- a.score(dapc2)
names(temp2)
temp2<- optim.a.score(dapc2)
```
optimun retained number of pcs 10
```{r}
temp2$pop.score
temp2$mean
```



```{r}
dapc3 <- dapc(new_Erigenia_gl)
```
n.pca= 10, n.da= 3
```{r}
temp3 <- a.score(dapc3)
names(temp3)
temp3<- optim.a.score(dapc3)
```
optimum ertained number of  pcs 6
```{r}
temp3$pop.score
temp3$mean
```




```{r}
dapc4 <- dapc(new_Erigenia_gl)
```
n.pc=6, n.da=3
```{r}
temp4 <- a.score(dapc4)
names(temp4)
temp4<- optim.a.score(dapc4)
```
optimum number retained pcs 5
```{r}
temp4$pop.score
temp4$mean
```

We chose dapc3 as the best, due to the fact PC choice levels off at 10 and doesn't change much when you go lower, more PCs (10 vs 5 or 6) will provide more useful data and better inform the plot


```{r}
loadingplot(dapc3$var.contr)
```
```{r}
summary(dapc3)
```
$assign.per.pop
    BraddockTrail        CedarCreek       PeachBottom      RaccoonCreek    RyersonStation 
                1                 1                 1                 1                 1 
       SafeHarbor SlipperyRockCreek       YorkFurnace 
                1                 1                 1 

$prior.grp.size

    BraddockTrail        CedarCreek       PeachBottom      RaccoonCreek    RyersonStation 
               15                15                20                15                15 
       SafeHarbor SlipperyRockCreek       YorkFurnace 
                9                15                14 

$post.grp.size

    BraddockTrail        CedarCreek       PeachBottom      RaccoonCreek    RyersonStation 
               15                15                20                15                15 
       SafeHarbor SlipperyRockCreek       YorkFurnace 
                9                15                14 



```{r}
scatter(dapc3, scree.da=FALSE, bg="white", pch=20, cell=0, cstar=0, col=myCol, solid=.4,
cex=3,clab=0, leg=TRUE, txt.leg=paste("Cluster",1:8))
```

```{r good fig}
myCol <- c("burlywood4","darkviolet","green","orange","red","dodgerblue3","black","gold2")
#pdf("DAPC.pdf", width = 6, height = 4)
scatter(dapc3, posi.da="topleft", bg="white",
pch=17:20, cex= 2, cstar=0, col=myCol, scree.pca=TRUE,
posi.pca="topright")
#dev.off()
```
Although this shows are east and west pops pretty close together, they're still split by the vert axis, so I don't hate it. 

```{r}
assignplot(dapc3)

```
3 is better so I think PC=10 is best

```{r}
save(dapc3, file="dapc3.RData")
```

```{r}
dapc3$assign
```
```{r}
dapc3$posterior
```
```{r}
loadingplot(dapc3$var.contr)
```


##STRUCTURE-like plot- not super effective could  come back to
```{r DAPC structure plot}
compoplot(dapc3,col = myCol, posi = 'bottom')
```
FUN!

##PCA 

###PCA with dartR
mostly to check the file for quality/make sure it works
```{r}
pc <- gl.pcoa(new_Erigenia_gl, nfactors=5)
```

```{r}
gl.pcoa.plot(pc, new_Erigenia_gl, labels="pop", xaxis=1, yaxis=2)
```
```{r}
gl.pcoa.plot(pc, new_Erigenia_gl, labels="pop", xaxis=2, yaxis=3)
```
```{r}
gl.pcoa.plot(pc, new_Erigenia_gl, labels="pop", xaxis=1, yaxis=3)
```


###PCA with adegenet
```{r} 
erigenia.pca <- glPca(new_Erigenia_gl, nf = 3)
barplot(100*erigenia.pca$eig/sum(erigenia.pca$eig), col = heat.colors(50), main="PCA Eigenvalues")
title(ylab="Percent of variance\nexplained", line = 2)
title(xlab="Eigenvalues", line = 1)
```


```{r saving glPCA with adegenet}
save(erigenia.pca, file= "erigenia.pca.RData")
```

```{r glPCA with ggplot}
erigenia.pca.scores <- as.data.frame(erigenia.pca$scores)
erigenia.pca.scores$pop <- pop(new_Erigenia_gl)
```



```{r}
set.seed(17)
p_1_2 <- ggplot(erigenia.pca.scores, aes(x=PC1, y=PC2, colour=pop)) 
p_1_2  <- p_1_2  + geom_point(size=2)
#p_1_2  <- p_1_2  + stat_ellipse(level = 0.95, size = 1) -this doesn't work with theses points, but could be used for better clustering 
p_1_2  <- p_1_2  + scale_color_manual(values = myCol) 
p_1_2  <- p_1_2  + geom_hline(yintercept = 0) 
p_1_2  <- p_1_2  + geom_vline(xintercept = 0) 
p_1_2  <- p_1_2  + theme_bw()

p_1_2
```
pc 1 =  24.6%
pc 2 = 5%

```{r other PCs in ggplot}
set.seed(17)
p_2_3 <- ggplot(erigenia.pca.scores, aes(x=PC2, y=PC3, colour=pop)) 
p_2_3 <- p_2_3 + geom_point(size=2)
#p_2_3 <- p_2_3 + stat_ellipse(level = 0.95, size = 1)
p_2_3 <- p_2_3 + scale_color_manual(values = myCol) 
p_2_3 <- p_2_3 + geom_hline(yintercept = 0) 
p_2_3 <- p_2_3 + geom_vline(xintercept = 0) 
p_2_3 <- p_2_3 + theme_bw()

p_2_3
```
not enough % described in PCs (see above)

```{r other PCs in ggplot 2}
set.seed(17)
p_1_3 <- ggplot(erigenia.pca.scores, aes(x=PC1, y=PC3, colour=pop)) 
p_1_3 <- p_1_3 + geom_point(size=2)
#p_1_3 <- p_1_3 + stat_ellipse(level = 0.95, size = 1)
p_1_3 <- p_1_3 + scale_color_manual(values = myCol) 
p_1_3 <- p_1_3 + geom_hline(yintercept = 0) 
p_1_3 <- p_1_3 + geom_vline(xintercept = 0) 
p_1_3 <- p_1_3 + theme_bw()

p_1_3
```

